p4-fusion/main.cc

```diff
@@ -63,9 +64,67 @@ int Main(int argc, char** argv)
 		return 1;
 	}
 
-	// Set the signal here because it gets reset after P4API library is initialized.
-	std::signal(SIGINT, SignalHandler);
-	std::signal(SIGTERM, SignalHandler);
+	// Block signals from being handled by the main thread, and all future threads.
+	//
+	// - SIGINT, SIGTERM, SIGHUP: thread pool will be shutdown and std::exit will be called
+	// - SIGUSR1: only sent by main() to tell the signal handler thread to exit
+	sigset_t blockedSignals;
+	sigemptyset(&blockedSignals);
+	for (auto sig : { SIGINT, SIGTERM, SIGHUP, SIGUSR1 })
+	{
+		sigaddset(&blockedSignals, sig);
+	}
+
+	int rc = pthread_sigmask(SIG_BLOCK, &blockedSignals, nullptr);
+	if (rc != 0)
+	{
+		ERR("(signal handler) failed to block signals: (" << errno << ") " << strerror(errno));
+		return 1;
+	}
+
+	// Create the thread pool
+	int networkThreads = arguments.GetNetworkThreads();
+	PRINT("Network Threads: " << networkThreads);
+	PRINT("Creating " << networkThreads << " network threads");
+	ThreadPool pool(networkThreads);
+	SUCCESS("Created " << pool.GetThreadCount() << " threads in thread pool");
+
+	// Spawn a thread to handle signals.
+	// The thread will block and wait for signals to arrive and then shutdown the thread pool, unless it receives SIGUSR1,
+	// in which case it will just exit (since main() is handling the shutdown).
+	//
+	// Using a separate thread for purely signal handling allows us to use non-reentrant functions
+	// (such as std::cout, condition variables, etc.) in the signal handler.
+	auto signalHandlingThread = std::thread(
+	    [&blockedSignals, &pool]()
+	    {
+		    // Wait for signals to arrive.
+		    int sig = 0;
+		    int rc = sigwait(&blockedSignals, &sig);
+		    if (rc != 0)
+		    {
+			    ERR("(signal handler) failed to wait for signals: (" << errno << ") " << strerror(errno));
+			    pool.ShutDown();
+			    std::exit(rc);
+		    }
+
+		    // Did main() tell us to shutdown?
+		    if (sig == SIGUSR1)
+		    {
+			    // Yes, so no need to print anything - just exit.
+			    return;
+		    }
+
+		    // Otherwise, we received a signal from the OS - print a message and shutdown.
+		    if (!sigismember(&blockedSignals, sig))
+		    {
+			    ERR("(signal handler): WARNING: received signal (" << sig << ") \"" << strsignal(sig) << "\" that is not blocked, this should not happen and indicates a logic error in the signal handler.");
+		    }
+
+		    ERR("(signal handler) received signal (" << sig << ") \"" << strsignal(sig) << "\", shutting down thread pool");
+		    pool.ShutDown();
```

Comment:
The main thread does a few more things on shutdown, like making sure tracing data is written, and that the P4 API is properly shut down. Can we somehow do the same here?

Reply:
From the main thread we're also hard-interrupting writing commits and trees now, since we call exit here.

I think in go I would finish writing the current changelist and the waitfordownload should be canceled with a context cancel error. Not sure how I would express that in C++.

Not a thing we have to solve in this PR, but feels a bit abrupt.

Reply:
> The main thread does a few more things on shutdown, like making sure tracing data is written, and that the P4 API is properly shut down. Can we somehow do the same here?

Done in [`43acb58` (#9)](https://github.com/sourcegraph/p4-fusion/pull/9/commits/43acb58f8c3beb1544d7d31621701478ba5c2de3)

Reply:
> From the main thread we're also hard-interrupting writing commits and trees now, since we call exit here.
> 
> I think in go I would finish writing the current changelist and the waitfordownload should be canceled with a context cancel error. Not sure how I would express that in C++.
> 
> Not a thing we have to solve in this PR, but feels a bit abrupt.

Maybe I'm missing something, but at a quick glance I would think this work would be encapsulated in a job (and will be cleanly shutdown when the thread pool is shutdown - but maybe I'm missing something)? We can chat on Monday about this though